miniapps/autodiff/example.hpp

```diff
@@ -0,0 +1,855 @@
+
+      TDataType norm2 = uu[0] * uu[0] + uu[1] * uu[1] + uu[2] * uu[2];
+      TDataType tvar = pow(ee * ee + norm2, (pp - 2.0) / 2.0);
+
+      rr[0] = tvar * uu[0];
+      rr[1] = tvar * uu[1];
+      rr[2] = tvar * uu[2];
+      rr[3] = -ff;
+   }
+};
+
+///Defines template class (functor) for evaluating the energy
+/// of the p-Laplacian problem. The input parameters vparam are:
+/// vparam[0] - the p-Laplacian power, vparam[1] small value
+/// ensuring exsitance of an unique solution, and vparam[2] -
+/// the distributed extenal input to the PDE.
+template<typename TDataType, typename TParamVector, typename TStateVector
+         , int state_size, int param_size>
+class MyQFunctor
+{
+public:
+   ///Returns the energy of a  p-Laplacian for state field input
+   /// provided in vector uu and parameters provided in vector
+   /// vparam.
+   TDataType operator()(TParamVector &vparam, TStateVector &uu)
+   {
+      MFEM_ASSERT(state_size==4,"MyQFunctor state_size should be equal to 4!");
+      MFEM_ASSERT(param_size==3,"MyQFunctor param_size should be equal to 3!");
+      double pp = vparam[0];
+      double ee = vparam[1];
+      double ff = vparam[2];
+
+      TDataType u = uu[3];
+      TDataType norm2 = uu[0] * uu[0] + uu[1] * uu[1] + uu[2] * uu[2];
+
+      TDataType rez = pow(ee * ee + norm2, pp / 2.0) / pp - ff * u;
+      return rez;
+   }
+};
+
+
+///Implements integrator for a p-Laplacian problem.
+/// The integrator is based on a class QFunction utilized for
+/// evaluating the energy, the first derivative (residual) and
+/// the Hessian of the energy (the Jacobian of the residual).
+template<class CQVectAutoDiff>
+class pLaplaceAD : public NonlinearFormIntegrator
+{
+protected:
+   Coefficient *pp;
+   Coefficient *coeff;
+   Coefficient *load;
+
+   CQVectAutoDiff rdf;
+
+public:
+   pLaplaceAD()
+   {
+      coeff = nullptr;
+      pp = nullptr;
+   }
+
+   pLaplaceAD(Coefficient &pp_) : pp(&pp_), coeff(nullptr), load(nullptr) {}
+
+   pLaplaceAD(Coefficient &pp_, Coefficient &q, Coefficient &ld_)
+      : pp(&pp_), coeff(&q), load(&ld_)
+   {}
+
+   virtual ~pLaplaceAD() {}
+
+   virtual double GetElementEnergy(const FiniteElement &el,
+                                   ElementTransformation &trans,
+                                   const Vector &elfun) override
+   {
+      double energy = 0.0;
+      int ndof = el.GetDof();
+      int ndim = el.GetDim();
+      int spaceDim = trans.GetSpaceDim();
+      bool square = (ndim == spaceDim);
+      const IntegrationRule *ir = NULL;
+      int order = 2 * el.GetOrder() + trans.OrderGrad(&el);
+      ir = &IntRules.Get(el.GetGeomType(), order);
+
+      Vector shapef(ndof);
+      DenseMatrix dshape_iso(ndof, ndim);
+      DenseMatrix dshape_xyz(ndof, spaceDim);
+      Vector grad(spaceDim);
+
+      Vector vparam(3); //[power, epsilon, load]
+      Vector uu(4);     //[diff_x,diff_y,diff_z,u]
+
+      uu = 0.0;
+      vparam[0] = 2.0;  //default power
```

Comment:
It would be more reusable if these were parameters instead of hard coded, but I understand that this is an example.

Reply:
They are parameter and users can provide coefficients- I just provide default values for them.  These are moved to the constructor in the modified code.